extension/app/ts/background/simulationModeHanders.ts

```diff
@@ -69,27 +69,47 @@ export async function sendTransaction(getActiveAddressForDomain: (websiteAccess:
 	return await openConfirmTransactionDialog(requestId, origin, simulationMode, formTransaction)
 }
 
-export async function call(simulator: Simulator, request: EthCallParams) {
+async function singleCallWithFromOverride(simulator: Simulator, request: EthCallParams, from: bigint) {
+	const callParams = request.params[0]
+	const blockTag = request.params.length > 1 ? request.params[1] : 'latest' as const
+	const input = callParams.data !== undefined ? callParams.data : new Uint8Array()
+	const gasPrice = callParams.gasPrice !== undefined ? callParams.gasPrice : 0n
+	const value = callParams.value !== undefined ? callParams.value : 0n
 	const transaction = {
 		type: '1559' as const,
-		from: defaultCallAddress,
+		from,
 		chainId: await simulator.ethereum.getChainId(),
-		nonce: await simulator.simulationModeNode.getTransactionCount(defaultCallAddress),
-		maxFeePerGas: 0n,
+		nonce: await simulator.simulationModeNode.getTransactionCount(from),
+		maxFeePerGas: gasPrice,
 		maxPriorityFeePerGas: 0n,
-		gas: await simulator.simulationModeNode.estimateGas({
-			from: defaultCallAddress,
-			to: request.params[0].to,
-			data: request.params[0].data
+		gas: callParams.gas !== undefined ? callParams.gas : await simulator.simulationModeNode.estimateGas({
+			from,
+			to: callParams.to,
+			data: input,
+			gasPrice,
+			value,
 		}),
-		to: request.params[0].to,
-		value: 0n,
-		input: 'data' in request.params[0] || request.params[0].data === undefined ? new Uint8Array() : request.params[0].data,
+		to: callParams.to,
+		value,
+		input,
 		accessList: [],
 	}
-	const result = request.params.length > 1 ? await simulator.simulationModeNode.call(transaction, request.params[1]) : await simulator.simulationModeNode.call(transaction)
-	return { result: result }
+	return await simulator.simulationModeNode.call(transaction, blockTag)
 }
+
+export async function call(simulator: Simulator, request: EthCallParams) {
+	const callParams = request.params[0]
+	const from = callParams.from !== undefined && !KNOWN_CONTRACT_CALLER_ADDRESSES.includes(callParams.from) ? callParams.from : defaultCallAddress
```

Comment:
so we can add addresses here to optimize against lots of failed calls. It is unfortunate that if a site does this over and over again (as it is likely to do) we will need to discover this over and over via initial failures, but it works for now.

Reply:
yeah, I was thinking of adding caching here, but I didn't want to make it overly complex. So that if it makes a call and identifies its contract, it would change the address always for those for that client